File: testSuite/test-parallelTreeBuild.pl
#!/usr/bin/env perl
use strict;
use warnings;
use PDL;
use PDL::IO::HDF5;
use PDL::NiceSlice;

# Run Galacticus models with and without parallelism in tree building.
# Andrew Benson (05-July-2023)

# Ensure output directory exists.
system("mkdir -p outputs");

# Run models.
system("export OMP_NUM_THREADS=1; cd ..; ./Galacticus.exe testSuite/parameters/parallelTreeBuildSerial.xml");
die("FAILED: failed to run serial model")
    unless ( $? == 0 );
system("export OMP_NUM_THREADS=4; cd ..; ./Galacticus.exe testSuite/parameters/parallelTreeBuildParallel.xml");
die("FAILED: failed to run parallel model")
    unless ( $? == 0 );

# Read model data.
my $data;
foreach my $modelName ( "parallelTreeBuildSerial", "parallelTreeBuildParallel" ) {
    my $model    = new PDL::IO::HDF5("outputs/".$modelName.".hdf5");
    my $analysis = $model->group("conditionalMassFunction");
    $data->{$modelName}->{$_} = $analysis->dataset($_)->get()
	foreach ( "conditionalMassFunction", "conditionalMassFunctionError" );
}

# Compare results.
my $selection       = which($data->{'parallelTreeBuildSerial'}->{'conditionalMassFunction'}->flat() > 0.0);
my $errorNormalized = abs($data->{'parallelTreeBuildParallel'}->{'conditionalMassFunction'}->flat()->($selection)-$data->{'parallelTreeBuildSerial'}->{'conditionalMassFunction'}->flat()->($selection))/$data->{'parallelTreeBuildSerial'}->{'conditionalMassFunctionError'}->flat()->($selection);
my $status          = any($errorNormalized > 4.0) ? "FAIL" : "SUCCESS";
print $status.": parallel tree build\n";
if ( $status eq "FAIL" ) {
    print "Conditional mass function:\n";
    print "\t  Serial build: ".$data->{'parallelTreeBuildSerial'  }->{'conditionalMassFunction'     }->flat()->($selection)."\n";
    print "\tParallel build: ".$data->{'parallelTreeBuildParallel'}->{'conditionalMassFunction'     }->flat()->($selection)."\n";
    print "\t   Uncertainty: ".$data->{'parallelTreeBuildSerial'  }->{'conditionalMassFunctionError'}->flat()->($selection)."\n";
}

exit;

File: testSuite/test-parallelTreeBuild.py
#!/usr/bin/env python3
import subprocess
import sys
import h5py
import numpy as np

# Run Galacticus models with and without parallelism in tree building.
# Andrew Benson (ported to Python)

# Ensure output directory exists.
subprocess.run("mkdir -p outputs", shell=True)

# Run models.
status = subprocess.run("export OMP_NUM_THREADS=1; cd ..; ./Galacticus.exe testSuite/parameters/parallelTreeBuildSerial.xml", shell=True)
if status.returncode != 0:
    print("FAILED: failed to run serial model")
    sys.exit(0)

status = subprocess.run("export OMP_NUM_THREADS=4; cd ..; ./Galacticus.exe testSuite/parameters/parallelTreeBuildParallel.xml", shell=True)
if status.returncode != 0:
    print("FAILED: failed to run parallel model")
    sys.exit(0)

# Read model data.
data = {}
for modelName in ("parallelTreeBuildSerial", "parallelTreeBuildParallel"):
    with h5py.File(f"outputs/{modelName}.hdf5", "r") as model:
        analysis = model["conditionalMassFunction"]
        data[modelName] = {
            "conditionalMassFunction":      analysis["conditionalMassFunction"][:],
            "conditionalMassFunctionError": analysis["conditionalMassFunctionError"][:],
        }

# Compare results.
serialFlat   = data["parallelTreeBuildSerial"]["conditionalMassFunction"].flat[:]
parallelFlat = data["parallelTreeBuildParallel"]["conditionalMassFunction"].flat[:]
errorFlat    = data["parallelTreeBuildSerial"]["conditionalMassFunctionError"].flat[:]
selection    = np.where(serialFlat > 0.0)[0]
errorNormalized = np.abs(parallelFlat[selection] - serialFlat[selection]) / errorFlat[selection]
status_str   = "FAIL" if np.any(errorNormalized > 4.0) else "SUCCESS"
print(f"{status_str}: parallel tree build")
if status_str == "FAIL":
    print("Conditional mass function:")
    print(f"\t  Serial build: {serialFlat[selection]}")
    print(f"\tParallel build: {parallelFlat[selection]}")
    print(f"\t   Uncertainty: {errorFlat[selection]}")
